Objects/regions.c

```diff
@@ -92,9 +114,265 @@ struct regionmetadata {
+
+    if (new_parent != 0) {
+        regionmetadata_inc_rc(new_parent);
+    }
+}
+
+static regionmetadata* regionmetadata_get_parent(regionmetadata* self) {
+    self = regionmetadata_get_representative_or_null(self);
+    if (!self) {
+        // Constant regions never have a parent
+        return NULL;
+    }
+
+    // Make sure we actually look at the real parent and not a region that
+    // we're merged into.
+    Py_region_ptr_t parent_raw = regionmetadata_get_representative(Py_region_ptr(self->parent));
+    assert(!IS_CONST_REGION(parent_raw) && "Constant regions can never be a parent.");
+    regionmetadata* parent = _Py_CAST(regionmetadata*, parent_raw);
+
+    // Here we set the parent to the representative we received.
+    // In most cases this will do nothing, but in others it can decrese
+    // future lookup times.
+    regionmetadata_set_parent(self, parent);
+
+    return _Py_CAST(regionmetadata*, parent);
+}
+#define regionmetadata_get_parent(self) \
+    regionmetadata_get_parent(_Py_CAST(regionmetadata*, self))
+
+static bool regionmetadata_has_parent(regionmetadata* self) {
+    return regionmetadata_get_parent(self) != NULL;
+}
+
+// This implementation merges `self` into `other`. This only works if
+// `self` is an orphan or if it has the same parent as `other`.
+//
+// FIXME: This implementation is currently incomplete. But it should be
+// sufficient for now.
+__attribute__((unused))
+static PyObject* regionmetadata_merge(regionmetadata* self, Py_region_ptr_t other) {
+    self = regionmetadata_get_representative_or_null(self);
+    if (!self) {
+        assert(false && "There should never be an attempt to merge constant regions");
+        return NULL;
+    }
+
+    other = regionmetadata_get_representative(other);
+
+    // If self doesn't have a parent we can merge directly
+    regionmetadata* self_parent = regionmetadata_get_parent(self);
+    if (self_parent == 0) {
+        goto merge;
+    }
+
+    // Constant regions never have a parent.
+    if (IS_CONST_REGION(other)) {
+        goto fail;
+    }
+
+    // If both regions have a parent, they need to be the same.
+    if (self_parent != regionmetadata_get_parent(other)) {
+        goto fail;
+    }
+
+    // Update the RC of the previous parent.
+    Py_region_ptr_t parent = Py_region_ptr(self->parent);
+    if (parent != 0 && !IS_CONST_REGION(parent)) {
+        regionmetadata_dec_rc(parent);
+    }
+
+merge:
+    if (!IS_CONST_REGION(other)) {
+        regionmetadata_inc_rc(other);
+    }
+
+    self->parent = Py_region_ptr_with_tags(other | Py_METADATA_MERGE_TAG);
+    // No decref, since this is a week reference. Otherwise we would get
+    // a cycle on a non GC'ed object and the bridge.
+    self->bridge = NULL;
+    Py_RETURN_NONE;
+
+fail:
+    throw_region_error(self->bridge, _Py_CAST(regionmetadata*, other)->bridge,
+                       "Unable to merge regions with differen parents", NULL);
+    return NULL;
+}
```

Comment:
Merging needs to increase the reference count of the new root.  The RC should represent the number of regions and objects that have this as its parent.

Reply:
The ref count of the root is increase in line 270. (Unless it's `_Py_DEFAULT_REGION` or `_Py_IMMUTABLE_REGION`) or do I misunderstand you?

Reply:
Sorry, reading too quickly.  But reading slower I have some more comments.

I would have expected this function to be very symmetric.  Why are the argument types different?  On line 239 you check for `!self` but on 253 you check `IS_CONST_REGION(other)`.

Is it okay for both regions to have the same parent?  That will break external uniqueness.